Source/Storages/Local/LocalKeyStorage.swift

```diff
@@ -39,33 +39,129 @@ import VirgilSDK
 
 internal class LocalKeyStorage {
     internal let identity: String
-    private let crypto: VirgilCrypto
+    internal let crypto: VirgilCrypto
+
+    private var keyPair: VirgilKeyPair?
     private let keychainStorage: KeychainStorage
+    private let options: KeychainQueryOptions
+
+#if os(iOS)
+    internal convenience init(identity: String,
+                              crypto: VirgilCrypto,
+                              keychainStorage: KeychainStorage,
+                              biometricProtection: Bool,
+                              biometricPromt: String? = nil) throws {
+        let options = KeychainQueryOptions()
+        options.biometricallyProtected = biometricProtection
+
+        if let biometricPromt = biometricPromt {
+            options.biometricPromt = biometricPromt
+        }
+
+        try self.init(identity: identity, crypto: crypto, keychainStorage: keychainStorage, options: options)
+    }
+
+    private var backupName: String {
+        return "E3KIT-BACKUP-" + self.identity
+    }
+
+    private func store(backup data: Data) throws {
+        _ = try self.keychainStorage.store(data: data,
+                                           withName: self.backupName,
+                                           meta: nil)
+    }
+
+    private func deleteBackup() throws {
+        try self.keychainStorage.deleteEntry(withName: self.backupName)
+    }
+
+    private func applyBackup() throws {
+        guard let data = try self.retrieve(name: self.backupName) else {
+            return
+        }
 
-    internal init(identity: String, crypto: VirgilCrypto, keychainStorage: KeychainStorage) {
+        try self.store(data: data)
+
+        try self.deleteBackup()
+    }
+
+    internal func setBiometricProtection(to set: Bool) throws {
+        guard self.options.biometricallyProtected != set, self.keyPair != nil else {
+            return
+        }
+
+        let data = try self.crypto.exportPrivateKey(self.getKeyPair().privateKey)
+
+        try self.store(backup: data)
+
+        try self.delete()
+        self.options.biometricallyProtected = set
+        try self.store(data: data)
+
+        try self.deleteBackup()
+    }
+#endif
+
+    internal required init(identity: String,
+                           crypto: VirgilCrypto,
+                           keychainStorage: KeychainStorage,
+                           options: KeychainQueryOptions = KeychainQueryOptions()) throws {
         self.identity = identity
         self.crypto = crypto
         self.keychainStorage = keychainStorage
+        self.options = options
+
+        try self.loadKeyPair()
```

Comment:
Load key-pair only when you need it

Reply:
User opens app, enters chat, tries to send new message and suddenly face id check appears? Doesn't seem to be good UX